### PR TITLE
sql: fix panic on UPDATE RETURNING * in middle of schema change

### LINKED ISSUE #32177: sentry: conn_executor.go:670: panic while executing 1 statements: UPDATE _ SET _ = IFNULL(_ + $1, $1), _ = IFNULL(_ + $2, $2), _ = IF((_ IS _) OR (_ < _), _, _) WHERE (((((_ = $3) AND (_ = $4)) AND (_ = $5)) AND (_ = $6)) AND (_ = $7)) AND (_ = _) LIMIT _ RETURNING *: caused by <redacted>
https://sentry.io/cockroach-labs/cockroachdb/issues/755922465/

conn_executor.go:670: panic while executing 1 statements: UPDATE _ SET _ = IFNULL(_ + $1, $1), _ = IFNULL(_ + $2, $2), _ = IF((_ IS _) OR (_ < _), _, _) WHERE (((((_ = $3) AND (_ = $4)) AND (_ = $5)) AND (_ = $6)) AND (_ = $7)) AND (_ = _) LIMIT _ RETURNING *: caused by <redacted>

github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1

> stacktrace: {u'frames': [{u'function': u'func4', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go', u'lineno': 309, u'in_app': True}, {u'function': u'ServeConn', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*Server)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go', u'lineno': 420, u'in_app': True}, {u'function': u'run', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go', u'lineno': 1200, u'in_app': True}, {u'function': u'execStmt', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go', u'lineno': 95, u'in_app': True}, {u'function': u'execStmtInOpenState', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go', u'lineno': 403, u'in_app': True}, {u'function': u'dispatchToExecutionEngine', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go', u'lineno': 825, u'in_app': True}, {u'function': u'execWithDistSQLEngine', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go', u'lineno': 983, u'in_app': True}, {u'function': u'PlanAndRun', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go', u'lineno': 746, u'in_app': True}, {u'function': u'Run', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go', u'lineno': 253, u'in_app': True}, {u'function': u'StartSync', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go', u'lineno': 602, u'in_app': True}, {u'function': u'Run', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go', u'lineno': 731, u'in_app': True}, {u'function': u'Run', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql/distsqlrun', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go', u'lineno': 170, u'in_app': True}, {u'function': u'Next', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go', u'lineno': 205, u'in_app': True}, {u'function': u'Next', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_batch.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*serializeNode)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/plan_batch.go', u'lineno': 123, u'in_app': True}, {u'function': u'BatchedNext', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/update.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*updateNode)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/update.go', u'lineno': 533, u'in_app': True}, {u'function': u'processSourceRow', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/update.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*updateNode)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/update.go', u'lineno': 678, u'in_app': True}, {u'function': u'AddRow', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer)', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go', u'lineno': 280, u'in_app': True}, {u'function': u'gopanic', u'abs_path': u'/usr/local/go/src/runtime/panic.go', u'module': u'runtime', u'filename': u'runtime/panic.go', u'lineno': 502, u'in_app': False}, {u'function': u'call32', u'abs_path': u'/usr/local/go/src/runtime/asm_amd64.s', u'module': u'runtime', u'filename': u'runtime/asm_amd64.s', u'lineno': 573, u'in_app': False}, {u'function': u'func1', u'abs_path': u'/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go', u'module': u'github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn', u'filename': u'github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go', u'lineno': 418, u'in_app': True}]}
type: *log.safeError
value: conn_executor.go:670: panic while executing 1 statements: UPDATE _ SET _ = IFNULL(_ + $1, $1), _ = IFNULL(_ + $2, $2), _ = IF((_ IS _) OR (_ < _), _, _) WHERE (((((_ = $3) AND (_ = $4)) AND (_ = $5)) AND (_ = $6)) AND (_ = $7)) AND (_ = _) LIMIT _ RETURNING *: caused by <redacted>

### PR DESCRIPTION
**Pull request by vivekmenezes**
fixes #32177

Release note (sql change): fix panic on UPDATE RETURNING * during
a schema change.

### CODE REVIEW

**Comment by vivekmenezes** [comment]
@jordanlewis added more tests including tests that show case issue #29494 which I've reopened

**Comment by vivekmenezes** [comment]
@jordanlewis since the customer has a way forward without this fix, I'm going to work on getting this fix in without reopening #29494

Hold off on reviewing this.

**Comment by vivekmenezes** [comment]
@jordanlewis @knz I was able to fix this without reverting the change. The fix turned out a lot simpler than I expected.

**Comment by vivekmenezes** [comment]
ping

File: pkg/sql/update.go [hunk 1]
```diff
@@ -676,7 +676,15 @@ func (u *updateNode) processSourceRow(params runParams, sourceVals tree.Datums)
...
 	// If result rows need to be accumulated, do it.
 	if u.run.rows != nil {
-		if _, err := u.run.rows.AddRow(params.ctx, newValues); err != nil {
+		// The new values can include all columns, the construction of the
+		// values has used publicAndNonPublicColumns so the values may
+		// contain additional columns for every newly added column not yet
+		// visible. We do not want them to be available for RETURNING.
+		//
+		// MakeUpdater guarantees that the first columns of the new values
+		// are those specified u.column .
```
Comment (knz) [inline pkg/sql/update.go]: nit: "those specified in u.columns."

**Review by knz — APPROVED** [review]
nice!

File: pkg/sql/update.go (hunk 1 above)
Comment (vivekmenezes) [reply pkg/sql/update.go]: Done

**Pull request merged**